### DISCUSSION TITLE
Loading sktime data by writing a .ts file and reading it into sktime

### DISCUSSION THREAD

**Discussion by MagnusNordboe** [alan-turing-institute/sktime#4260] [Q&A] [answered]
**Most likely I'm being a big dummy and simply misunderstanding how to write the ts file format**
**For some reason, github auto formats "data" in my tsfile to "DaTa". Dunno why, it's lowercase in my file.**

I'm doing a master's project where I try to analyze large amounts of Prometheus data, stored in csv files. Those files are quite large, so for now I'm working with two dummy files containing some data in the same format.

The problem is that I get the error `too many values to unpack (expected 2)` whenever I supply more than 2 data points. 

Here's a ts file I tried writing with some dummy data:
`    

@problemName sockshop
@timeStamps false
@missing false
@univariate false
@dimensions 3
@equalLength true
@seriesLength 8
@classLabel false
@data
312,345,355,355,356,354,355,356:0,0,0,0,1,1,1,2:45,87,23,67,12,3,654,4`

This raises `too many values to unpack (expected 2)` when trying load it with load_from_tsfile()

In my actual data, I would have several newlines for other csv files as well. Each of my csv files contain ~40 observations over ~200 dimensions. I have ~50 of those files. In my understanding of the ts file format, this would mean a ts file with ~50 newlines, each line having ~200 colons with ~40 values separated by commas within those colons. 

Here's a file I did get to load into the sktime format:
`    

@problemName sockshop
@timeStamps false
@missing false
@univariate false
@dimensions 3
@equalLength true
@seriesLength 8
@classLabel false
@data
312:313
13:12
31:33
`

output: 
`('dim_0', 'dim_1')`
The type here is tuple, and it contains no more information, i.e. none of the data I passed into it.

Here's a version that crashes with the same  `too many values to unpack (expected 2)` error:

`     

@problemName sockshop
@timeStamps false
@missing false
@univariate false
@dimensions 3
@equalLength true
@seriesLength 8
@classLabel false
@data
312:313:314
13:12:12
31:33:33
`

The same happens with comma separation instead of colon separation. 

Check out the simple dummy csv files I'm trying to add to ts. Step one is just importing one, but I want to combine both of them in a single ts file (and for the real implementation, many more that just 2)

[simple1.csv](https://github.com/sktime/sktime/files/10807836/simple1.csv)
[simple2.csv](https://github.com/sktime/sktime/files/10807846/simple2.csv)

**Comment by fkiraly** [comment]
uh-oh - I suppose since GitHub tags also start with an `@` symbol, you just tagged a lot of unrelated people. That's also why it capitalizes data as DaTa, it's the name of a GitHub user.

I hope everyone can see this with a pinch of humour :-)
Welcome to sktime

**Reply by fkiraly** [reply]
for everyone tagged: we are discussing a data format where reserved fields start with an `@`, and GitHub parsed these as your GitHub names

**Comment by fkiraly** [comment, marked as answer]
can you post the full code that gives you the error?

**Reply by MagnusNordboe** [reply]
Out of respect for all the poor github users, I will just paste the python file as a txt: it's quite short.
[to_ts.py.txt](https://github.com/sktime/sktime/files/10808651/to_ts.py.txt)

**Reply by fkiraly** [reply]
well, the python file should not be a problem as it does not contain the `@` symbol.
Could you just kindly paste it in a code block (three apostrophes delimiting)?

**Reply by MagnusNordboe** [reply]
`
with open("output.ts", mode="w") as file:
    file.writelines(['A bunch of @ symbols for manually adding some tags, removed for this apste])

    with open("../Test_datasets/simple1.csv") as csv:
        #calling readline seems the simplest way to skip the first line, which contains timestamps. Perhaps use this line to add timestamps?
        csv.readline()
        #for the very first line of data, we do separately just because the first one doesnt begin with a colon
        #This code looks messy but does the exact same as the more readable code below, except adding a colon.
        file.write(csv.readline().replace("\n","").split(",",1)[1])
        for x in csv:
            
            #Remove the newline at the end of each line, and split the identifier form the string so we only have the numbers
            identifier, numberlist = x.replace("\n","").split(",",1)
            file.write(":" + numberlist)
        
    
test = load_from_tsfile("testing.ts")
print(test)
type(test)
print(type(test[0]))
reading = open("testing.ts", mode="r")
print(reading.read())
`

**Reply by fkiraly** [reply]
I see, this is the code where you write to the file.

I meant the code before that, where you read the `csv` and it does not fit an `sktime` format.
That would be the best place to fix the formatting, rather than trying an export/import workaround

**Reply by MagnusNordboe** [reply]
Alright, I based my code on a resource I found online about converting csvs to sktime format. 
[Convert_CSV.zip](https://github.com/sktime/sktime/files/10822268/Convert_CSV.zip)


It kept throwing errors about inputs of object dtype. You can see a long, messy trail of my frustrated attempts to fix this to no avail.

**Reply by fkiraly** [reply]
I've looked into the notebook, I would recommend you use the multiindex format and not the nested format.
The nested format has not been working properly since `pandas` updated to 1.5.0, nested objects are not officially supported so we have soft-deprecated it (using it less, highlighting it less) and might properly deprecate it later.

The classification tutorial still relies on it, and should be updated at some point (any help appreciated as `sktime` is volunteering driven!)

To get an example of the multiindex format, type

```python
from sktime.datatypes import get_examples

X = get_examples("pd-multiindex")[0]
# change the 0 to other numbers for more examples
```

For an explanation of the supported in-memory data formats for inputs, look at the notebook
`examples/AA_datatypes_and_datasets.ipynb`
e.g., here on binder: https://mybinder.org/v2/gh/sktime/sktime/main?filepath=examples

**Reply by fkiraly** [reply]
FYI @MagnusNordboe, let me know whether this helps - short summary, try to avoid the nested format as it is is soft-deprecated and can gives problems on newer `pandas` versions if you want to create it manually.

**Reply by MagnusNordboe** [reply]
Hi again!

Thanks a lot, changing to multiindex implementation fixed things for me (with quite a few headaches on the way, but we got there).

The sktime.datatypes.check_raise() function and its lacking documentation threw me quite a loop there. Only after running head first into it for quite a while did I dig through the code file for the function and track down the datatypes.MTYPE_REGISTER to find that multiindex was a separate thing to check for (I was running it with mtype=pd.DataFrame the whole time). Might make a pull request on that, or submit a github issue for it.

I am looking to run a code like the one included below on many csv files, each file representing an instance in the finished MTS. I'm still a beginner at Pandas. Do you see some elegant way for me to expand upon the dataframe when running on the subsequent simple2.csv file?

[convert_csv_panel.zip](https://github.com/sktime/sktime/files/10844111/convert_csv_panel.zip)

**Reply by fkiraly** [reply]
would having a formatting routine, collecting inputs from files, and then having `pd.concat` work? Using the `keys` argument if you need an additional level of hierarchy.

**Comment by fkiraly** [comment]
> The sktime.datatypes.check_raise() function and its lacking documentation threw me quite a loop there. Only after running head first into it for quite a while did I dig through the code file for the function and track down the datatypes.MTYPE_REGISTER to find that multiindex was a separate thing to check for (I was running it with mtype=pd.DataFrame the whole time). Might make a pull request on that, or submit a github issue for it.

That would be much appreciated!

Realldy sorry that the experience wasn't too great, I'm aware of the misleading naming. This is dating back to a time where there were no hierarchical data type, and the only `DataFrame` based format was `pd.DataFrame`.
